### ISSUE TITLE
Course: Generative AI and LLMs: Architecture and Data Preparation-Module 2

### ISSUE THREAD

**Issue by niyontezejado** [niyontezejado/GenAI#2]
# MODULE 2: DATA PREPARATION FOR LLMs

## 1. Objectives

- Tokenization process, tokenization methods, and the use of tokenizers
- Implement tokenization
- Explain how data loaders are used to prepare data for LLMs
- Create an NLP loader

## 2. Tokenization

Tokenization is the breaking of sentence into smaller pieces. This makes the model to understand the text better.

Example: IBM taught me tokenization
Tokens: ['IMB', 'taught', 'me', 'tokenization']

- Tokenizer generates tokens (NLTK, and spaCy)
- Tokenization methods: word-based, character-based, and subword-based
- Advantage of word-based tokenization:
    - preserves semantic meaning
- Desadvantage of word-based tokenization:
    - increases the model's performance

- Algorithms for subword-based tokenization: WordPiece, Unigram, SentencePiece

## 3. Overview of Data Loaders

- **Data loaders** : helps prepare and load data, implemented using DataLoader class in PyTorch.
- Purpose of using NLP data loader: efficient loading and preprocessing of textual data: 
 efficient batching and shuffling of data, memory optimization through on-the-fly preprocessing, seamless integration with PyTorch training pipeline, simplified data augmentation and preprocessing.

- PyTorch and TensorFlow have a dedicated Data Loader class.
